spec/PeriodicBackgroundSync-index.bs

```diff
@@ -92,3 +94,95 @@ Fetch requests within the <code>periodicsync</code> event while [=in the backgro
+
+Algorithms {#algorithms}
+=========================
+## <dfn>Calculate a time to fire</dfn> ## {#caculate-time-to-fire}
+This section describes how a user agent can calculate the [=time to fire=] for a [=periodicsync registration=], <var>registration</var>. The time interval between periodicSync events for |registration| MUST be greater than or equal to the [=minInterval=] value associated with |registration|.
+
+The user agent MAY include factors such as user engagement with the origin to decide this time interval, allowing origins with high user engagement to update their web apps more often. In addition, the user agent SHOULD have per-origin and cross-origin caps on this interval, as recommended in [[#privacy]] and Resource Usage sections.
+
+The user agent may also decide to retry each failed periodicSync event, capped to a user agent defined <dfn>maximum number of retries</dfn>.
+
+A possible algorithm to calculate the [=time to fire=], <var>timeToFire</var> for |registration| would involve running these steps:
+1. If this is the <dfn>first attempt of firing</dfn> the periodicSync event:
+    1. Let <var>now</var> represent the current time.
+    1. Let <var>siteEngagementLevel</var> represent the level of engagement the user has with the origin. Calculate <var>siteEngagementPenalty</var>, which is a factor of |siteEngagementLevel|, and is inversely proportional to it. Let <var>minPeriodicSyncInterval</var> be a global cap on time intervals between periodicSync events. Define <var>effectiveGap</var> as |siteEngagementPenalty|*|minPeriodicSyncInterval|.
+    1. Let <var>minIntervalForOrigin</var> be the minimum time interval allowed by the user agent, per origin. Set <var>delayForOrigin</var> to the multiple of |effectiveGap| greater than or equal to |minIntervalForOrigin|.
+    1. Let <var>timeTillScheduledEventForOrigin</var> be the time when the next periodicSync event is scheduled for the origin, if any, null otherwise.
+    1. If |timeTillScheduledEventForOrigin| is not null:
+        1. If |timeTillScheduledEventForOrigin| - |delayForOrigin| is greater than or equal to |minIntervalForOrigin|, abort these substeps.
+        1. If |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin| and abort these substeps.
+        1. If |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|.
+    1. Let <var>minIntervalAcrossOrigins</var> be the minimum interval allowed between periodicSync events across origins. Let <var>timeSinceLastPeriodicSync</var> be the time since a periodicSync event was fired for any origin, if any, null otherwise.
```

Comment:
timeSinceLastPeriodicSync is a bit handwavy, I don't think that's a big issue, but let's keep that in mind.

Reply:
What would you suggest instead?

Reply:
Something similar to the definition here, where you define the variable and how it's updated: https://www.w3.org/TR/service-workers-1/#service-worker-registration-lifetime

Reply:
I have defined a "time of last fire" for each periodicsync registration.
From my very little experience however, providing pseudocode which leaves no doubt around how a user agent might implement an algorithm is good enough, details like these are quite unambiguous and easy to implement.

Reply:
UPDATE: I realized this is a global variable, and so I don't see much added value in defining it and updating it when we fire the event, only to be able to refer it here, when we can easily describe what it means here. I've reverted to original since you didn't feel strongly about this, but let me know if you disagree, thanks.